File: Android/Pulsar/src/test/java/com/swmansion/pulsar/types/PatternSeekTest.kt
package com.swmansion.pulsar.types

import org.junit.Assert.assertEquals
import org.junit.Assert.assertSame
import org.junit.Assert.assertTrue
import org.junit.Test

class PatternSeekTest {

    private val ramp = PatternData(
        continuousPattern = ContinuousPattern(
            amplitude = listOf(ValuePoint(0L, 0f), ValuePoint(1000L, 1f)),
            frequency = listOf(ValuePoint(0L, 0.2f), ValuePoint(500L, 0.8f)),
        ),
        discretePattern = listOf(
            ConfigPoint(0L, 1f, 0.5f),
            ConfigPoint(400L, 0.8f, 0.4f),
            ConfigPoint(1000L, 0.6f, 0.3f),
        ),
    )

    @Test
    fun `duration is the last timestamp across both lines`() {
        assertEquals(1000L, PatternSeek.lastTimestampOf(ramp))
        val empty = PatternData(ContinuousPattern(emptyList(), emptyList()), emptyList())
        assertEquals(0L, PatternSeek.lastTimestampOf(empty))
    }

    @Test
    fun `seeking to zero returns the same pattern`() {
        assertSame(ramp, PatternSeek.patternFrom(ramp, 0L))
        assertSame(ramp, PatternSeek.patternFrom(ramp, -100L))
    }

    @Test
    fun `discrete events before the seek are dropped and the rest rebased`() {
        val seeked = PatternSeek.patternFrom(ramp, 400L)
        assertEquals(listOf(0L, 600L), seeked.discretePattern.map { it.time })
        assertEquals(listOf(0.8f, 0.6f), seeked.discretePattern.map { it.amplitude })
    }

    @Test
    fun `envelope is re-anchored on its interpolated value`() {
        val seeked = PatternSeek.patternFrom(ramp, 250L)
        assertEquals(listOf(0L, 750L), seeked.continuousPattern.amplitude.map { it.time })
        assertEquals(listOf(0.25f, 1f), seeked.continuousPattern.amplitude.map { it.value })
    }

    @Test
    fun `an envelope entirely before the seek holds its last value`() {
        val seeked = PatternSeek.patternFrom(ramp, 800L)
        assertEquals(listOf(0L, 200L), seeked.continuousPattern.frequency.map { it.time })
        assertEquals(listOf(0.8f, 0.8f), seeked.continuousPattern.frequency.map { it.value })
        assertTrue(seeked.continuousPattern.amplitude.isNotEmpty())
    }

    @Test
    fun `a held envelope collapses to one point once nothing remains`() {
        val seeked = PatternSeek.patternFrom(ramp, 1000L)
        assertEquals(listOf(ValuePoint(0L, 0.8f)), seeked.continuousPattern.frequency)
    }

    @Test
    fun `an empty envelope stays empty`() {
        val noFrequency = PatternData(
            ContinuousPattern(ramp.continuousPattern.amplitude, emptyList()),
            emptyList(),
        )
        assertTrue(PatternSeek.patternFrom(noFrequency, 250L).continuousPattern.frequency.isEmpty())
    }

    @Test
    fun `sound seeks into the file by the same amount`() {
        val seeked = PatternSeek.soundFrom(SoundData(uri = "clip.wav"), 300L)
        assertEquals(0L, seeked.offset)
        assertEquals(300L, seeked.startMs)
        assertEquals(0L, seeked.durationMs)
    }

    @Test
    fun `sound eats into the lead-in before it touches the file`() {
        val early = PatternSeek.soundFrom(SoundData(uri = "clip.wav", offset = 500L), 200L)
        assertEquals(300L, early.offset)
        assertEquals(0L, early.startMs)

        val late = PatternSeek.soundFrom(SoundData(uri = "clip.wav", offset = 500L), 800L)
        assertEquals(0L, late.offset)
        assertEquals(300L, late.startMs)
    }

    @Test
    fun `an authored trim window shrinks and its start advances`() {
        val sound = SoundData(uri = "clip.wav", startMs = 1000L, durationMs = 900L)
        val seeked = PatternSeek.soundFrom(sound, 400L)
        assertEquals(1400L, seeked.startMs)
        assertEquals(500L, seeked.durationMs)
    }
}

File: react-native/react-native-pulsar/src/__tests__/patternSeek.test.ts
import { patternDurationMs, patternFrom } from '../patternSeek';
import type { Pattern } from '../types';

const ramp: Pattern = {
  discretePattern: [
    { time: 0, amplitude: 1, frequency: 0.5 },
    { time: 400, amplitude: 0.8, frequency: 0.4 },
    { time: 1000, amplitude: 0.6, frequency: 0.3 },
  ],
  continuousPattern: {
    amplitude: [
      { time: 0, value: 0 },
      { time: 1000, value: 1 },
    ],
    frequency: [
      { time: 0, value: 0.2 },
      { time: 500, value: 0.8 },
    ],
  },
};

describe('patternDurationMs', () => {
  it('takes the last authored timestamp across both lines', () => {
    expect(patternDurationMs(ramp)).toBe(1000);
  });

  it('is zero for an empty pattern', () => {
    expect(
      patternDurationMs({
        discretePattern: [],
        continuousPattern: { amplitude: [], frequency: [] },
      })
    ).toBe(0);
  });
});

describe('patternFrom', () => {
  it('returns the pattern untouched at or before zero', () => {
    expect(patternFrom(ramp, 0)).toBe(ramp);
    expect(patternFrom(ramp, -100)).toBe(ramp);
  });

  it('drops the discrete events before the seek and rebases the rest', () => {
    expect(patternFrom(ramp, 400).discretePattern).toEqual([
      { time: 0, amplitude: 0.8, frequency: 0.4 },
      { time: 600, amplitude: 0.6, frequency: 0.3 },
    ]);
  });

  it('re-anchors an envelope onto its interpolated value at the seek', () => {
    expect(patternFrom(ramp, 250).continuousPattern.amplitude).toEqual([
      { time: 0, value: 0.25 },
      { time: 750, value: 1 },
    ]);
  });

  it('holds the last value when every point sits before the seek', () => {
    // Emptying it instead would silence BOTH channels, since the composer only builds
    // the continuous line when amplitude and frequency are each non-empty.
    expect(patternFrom(ramp, 800).continuousPattern.frequency).toEqual([
      { time: 0, value: 0.8 },
      { time: 200, value: 0.8 },
    ]);
    expect(patternFrom(ramp, 800).continuousPattern.amplitude).not.toEqual([]);
  });

  it('collapses a held envelope to a single point once nothing remains', () => {
    expect(patternFrom(ramp, 1000).continuousPattern.frequency).toEqual([
      { time: 0, value: 0.8 },
    ]);
  });

  it('leaves an empty envelope empty', () => {
    const noFrequency: Pattern = {
      ...ramp,
      continuousPattern: { ...ramp.continuousPattern, frequency: [] },
    };
    expect(patternFrom(noFrequency, 250).continuousPattern.frequency).toEqual(
      []
    );
  });

  describe('sound', () => {
    const withSound = (sound: NonNullable<Pattern['sound']>): Pattern => ({
      ...ramp,
      sound,
    });

    it('seeks into the file by the same amount', () => {
      expect(patternFrom(withSound({ uri: 'clip.wav' }), 300).sound).toEqual({
        uri: 'clip.wav',
        offset: 0,
        start: 300,
        duration: 0,
      });
    });

    it('eats into the lead-in before it touches the file', () => {
      // 200ms into a 500ms lead-in: the audio has not begun, so only the wait shortens.
      expect(
        patternFrom(withSound({ uri: 'clip.wav', offset: 500 }), 200).sound
      ).toEqual({ uri: 'clip.wav', offset: 300, start: 0, duration: 0 });

      // Past the lead-in, the remainder is a seek into the file.
      expect(
        patternFrom(withSound({ uri: 'clip.wav', offset: 500 }), 800).sound
      ).toEqual({ uri: 'clip.wav', offset: 0, start: 300, duration: 0 });
    });

    it('shrinks an authored trim window and adds to its start', () => {
      expect(
        patternFrom(
          withSound({ uri: 'clip.wav', start: 1000, duration: 900 }),
          400
        ).sound
      ).toEqual({ uri: 'clip.wav', offset: 0, start: 1400, duration: 500 });
    });
  });
});
